### DISCUSSION TITLE
Is it possible to create a variable in a process whose scope is bounded by that process?

### DISCUSSION THREAD

**Discussion by dphiggs01** [nextflow-io/nextflow#4261] [Q&A] [answered]
Hello,
I am relatively new to Nextflow, so I apologize for the newbie question; I could not find the answer in the docs.

I have an example process (defined below), and it works correctly.
However, I would like to improve the readability by creating a variable based on an input parameter with scope for the entire process.

e.g., I would like to make the variable `file_name_prefix` visible for the entire process.

I would appreciate any input on best practices for addressing this type of challenge.

Thanks!

```groovy
//THIS WORKS
process FASTQC_SINGLE {
    tag "FASTQC on ${reads.getName().split("\\.")[0]}"
    container 'fastqc-multiqc:1.0.1'
    publishDir params.outdir, mode:'copy'

    input:
    path reads

    output:
    path "fastqc_${reads.getName().split("\\.")[0]}_logs" 

    script:
    def file_name_prefix = reads.getName().split("\\.")[0]
    """
    mkdir fastqc_${file_name_prefix}_logs
    fastqc -o fastqc_${file_name_prefix}_logs -f fastq -q ${reads}
    """
}
```

```groovy
//THIS DOES NOT WORK (BUT READS BETTER)
process FASTQC_SINGLE {
    tag "FASTQC on ${file_name_prefix}"
    container 'fastqc-multiqc:1.0.1'
    publishDir params.outdir, mode:'copy'

    input:
    path reads
    def file_name_prefix = reads.getName().split("\\.")[0]

    output:
    path "fastqc_${file_name_prefix}_logs" 

    script:
     """
    mkdir fastqc_${file_name_prefix}_logs
    fastqc -o fastqc_${file_name_prefix}_logs -f fastq -q ${reads}
    """
}
```

**Comment by bentsherman** [comment, marked as answer]
As far as I know, it is not possible to declare a local variable that is accessible to directives and outputs. But, the process syntax has a lot of grey area and there might be some hack that works. I would love to be proven wrong and turn it into a best practice 😄

**Reply by dphiggs01** [reply]
Thanks Ben!

**Comment by pditommaso** [comment]
The correct syntax is the first one.

**Reply by dphiggs01** [reply]
Thank You.
